src/utp_stream.cpp

```diff
@@ -963,6 +963,10 @@ void utp_socket_impl::writable()
 	if (should_delete()) return;
 
 	while(send_pkt());
+	// if the socket stalled while sending an ack, then send it
+	// now if needed
+	if (m_deferred_ack && !m_stalled)
+		send_ack();
```

Comment:
maybe I'm misunderstanding something here. but if the `send_pkt()`-loop above ended up sending *any* packets, we don't need to send a dedicated ACK packet. Since the ACK sequence number is always included in the header.

I would have expected this part of the patch to instead do something like: if `m_deferred_ack` is set, make the first call to `send_pkt()` be called with the `pkt_ack` flag. If it returns true, then move on to run the loop that attempts to drain the send buffer.

looking closer i see that `m_deferred_ack` is actually cleared by `send_pkt()` if it ends up sending a packet, so I suppose this also works. However, if `send_pkt()` doesn't send any packets (and doesn't clear `m_deferred_ack`), there shouldn't be a way for `m_stalled` to have been set either.

I'm leaning towards thinking that this code might be slightly simpler if it doesn't rely on`m_deferred_ack` and `m_stalled` being updated by the call to `send_pkt()`.